### PR TITLE
Release Notes proposal

### PR DESCRIPTION
**Pull request by david-mcmahon**
Get the conversation started.
cc @eparis @vishh @davidopp @thockin @dchen1107 
Ref #23070

### CODE REVIEW

**Comment by david-mcmahon** [comment]
Ref #23070

File: docs/proposals/release-notes.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+v1.2.0 demonstrated the need, at least for major releases like v1.2.0, for
+several sections in the published release notes.
+In order to provide a basic layout for release notes in the future,
+new releases can bootstrap [kubernetes/CHANGELOG.md](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md) with a template such as:
+
+```
+Major Themes
+* Add to or delete this section
+
+Other notable improvements)
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: delete )

File: docs/proposals/release-notes.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+Experimental Features
+* Add to or delete this section
+
+Action Required
+* Add to or delete this section
+
+Known Issues
+* Add to or delete this section
+
+Provider Specific Notes
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: Provider-specific

File: docs/proposals/release-notes.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+  that capture the release-visiblity intention of that PR with a field inserted
+  into the commit message:
+
+  ```
+  # **Should this commit be a RELEASE NOTE? Choose one or DELETE the line.
+  # **Specify "breaking" for flag changes and the like
+  RELEASE NOTE: std | breaking
+  ```
+
+1. This field in the commit message acts as a trigger during the process of
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: I was thinking we'd extend release-note-breaking-change to specify the section in which the release note should appear.

File: docs/proposals/release-notes.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+Action Required
+* Add to or delete this section
+
+Known Issues
+* Add to or delete this section
+
+Provider Specific Notes
+* Add to or delete this section
+
+Changelog
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: Other notable changes

File: docs/proposals/release-notes.md (hunk 3 above)
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: release-note-major-theme, release-note-improvement, release-note-experimental, release-note-action, release-note-issue, release-note-provider-{gce,aws,...}, release-note-other

Comment (david-mcmahon) [inline docs/proposals/release-notes.md]: Some thoughts:
* Can every one of those sections be auto-generated from PR headings?  Or will some of those sections be more likely free-form text authored by the release owner(s)?
* Will every alpha, beta and minor release have those sections?  My thought is that only major (1.1, 1.2) will have all of those.
* release note section fatigue.  We want to introduce something to the workflow that is easy/intuitive to use.  We can include guidance in the form of commit message comments, but we are working with limited space there and should aim for brevity / intuitive design anyway.  The tyranny of choice.
* Maybe some but not all of the suggested labels/types here?

**Comment by thockin** [comment]
It's pretty clear that not everyone installs the triggers in git.  Do we really want release notes to be part of the COMMIT messages or part of the PR?  GitHub now has PR templates - can we use that instead?

https://github.com/blog/2111-issue-and-pull-request-templates

**Comment by david-mcmahon** [comment]
OTOH, maybe we can simplify this even further.  Forgive me as I'm not 100% fluent in every handy tool available in our toolbox.   I see the k8-merge-robot removing labels.  How does one make use of this bot?  

We could scan for labels, like: ` release-note`, `release-note-breaking-change`, `release-note-whatever` and most importantly `release-note-NONE`.  Unless a `release-note-*` label is placed on the PR, no LGTM for you.  

This works closely to the existing workflow and makes the release note decision (and editing) in the hands of both the reviewer and the author (or anyone else viewing the PR).

So is this possible? If so I can update the doc with this.  I think this is our best solution with the least friction to the development/PR workflow.

**Comment by david-mcmahon** [comment]
Proposal updated with munger option.  All considered options listed and plan to move ahead with munger/bot and labels.

File: docs/proposals/release-notes.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+inconsistent release notes often with errors and omissions.
+
+### The (general) Solution
+
+Like most things in the development/release pipeline, the earlier you do it,
+the easier it is for everyone and the better the outcome.  Gather your release
+notes earlier in the development cycle, at the time the features and fixes are
+added.
+
+This concept was first implemented at Google for the Borg team to
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: We could omit this paragraph.

File: docs/proposals/release-notes.md [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+## Collection Design
+
+The munger/bot option fits most cleanly into the existing workflow.
+
+The design will include:
+
+1. New labels added to github: `release-note-NONE`, maybe others for new release note categories - see Layout section below
+1. A [new munger](https://github.com/kubernetes/kubernetes/issues/23409) that will:
+  * Initiate a `needs-release-note` label on all new PRs
+  * Disallow/Remove the `lgtm` label on all PRs labeled as `needs-release-note`
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: Rather than remove lgtm, we could simply block merge by the submit queue, as with ok-to-merge.

File: docs/proposals/release-notes.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+  * Auto-remove `needs-release-note` when one of the release-note-\* labels is added
+  * Special case for cherry-picked/branch PRs, release-note-NONE is not allowed
+
+## Publishing Design
+
+### Location
+
+With v1.2.0, the release notes were moved from their previous [github releases](https://github.com/kubernetes/kubernetes/releases)
+location to [kubernetes/CHANGELOG.md](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md).  Going forward this seems like a good plan.
+Other projects do similar.
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: similarly

File: docs/proposals/release-notes.md [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+The kuberntes.tar.gz download link is also displayed along with the release notes
+in [kubernetes/CHANGELOG.md](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md).
+
+Is there any reason to continue publishing anything to github releases if
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: Is there a reason not to? Also, why is that relevant to this proposal?

File: docs/proposals/release-notes.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+the complete release story is published in [kubernetes/CHANGELOG.md](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md)?
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: In general, we prefer relative links within the repo.

File: docs/proposals/release-notes.md [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+* Add to or delete this section
+
+Known Issues
+* Add to or delete this section
+
+Provider-specific Notes
+* Add to or delete this section
+
+Other notable changes
+* release-note and release-note-breaking-change PR titles
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: release-note-breaking-change should go into Action Required. I'd be in favor of changing that label name to reflect that.

File: docs/proposals/release-notes.md [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+  * Time estimate: ???
+* Updates to the tool that collects, organizes, publishes and sends release
+  notifications.
+  * Owner: @david-mcmahon
+  * Time estimate: A few days
+
+
+## Caveats / Considerations
+
+* As part of the planning and development workflow how can we capture
```
Comment (bgrant0607) [inline docs/proposals/release-notes.md]: They should probably go into the PR that enables the feature by default.

File: docs/proposals/release-notes.md (hunk 5 above)
Comment (david-mcmahon) [inline docs/proposals/release-notes.md]: Agreed.  It was originally in there to show some successful precedent in case the whole idea was not well received.

File: docs/proposals/release-notes.md (hunk 8 above)
Comment (david-mcmahon) [inline docs/proposals/release-notes.md]: If the CHANGELOG contains the full release notes including links to the binary(s), what content do we want to have in the github releases?  Maybe just a link to CHANGELOG?

File: docs/proposals/release-notes.md (hunk 10 above)
Comment (david-mcmahon) [inline docs/proposals/release-notes.md]: OK, updating that to release-note-action-required (from release-note-breaking-change)
cc @eparis

**Comment by david-mcmahon** [comment]
PTAL

File: docs/proposals/release-notes.md (hunk 6 above)
Comment (eparis) [inline docs/proposals/release-notes.md]: I'm worried we have lots of reviewers who look for assigned issues without LGTM but not many who would filter on do-not-merge in this case. The author of the PR, who would likely see the do-no-merge label would often be unable to set the correct rel-not label (not all PR authors have label/commit privs). But we know that anyone who could set the LGTM will also have to power to set/handle the rel-note issues. So I think the rel-note labeling should be enforced by the reviewer.

Comment (david-mcmahon) [inline docs/proposals/release-notes.md]: Both of these suggestions seem reasonable, but I would defer to those more familiar with the typical workflow here.  I'm sure we can't please all the people all the time, but getting it working for the majority use case should be the target.  Anyone else want to chime in here?

**Comment by eparis** [comment]
As soon as this merges we should merge https://github.com/kubernetes/contrib/pull/670 and restart the queue.

**Comment by david-mcmahon** [comment]
@bgrant0607 is this good for now?  We can update it if/when we add or modify release-note-* labels in the future.  But as an overall process, this doc covers the basics.  There are a number of PRs waiting to go in based on this including the release-note-* munger itself.

**Comment by eparis** [comment]
We can fix it as we go, LGTM
